compiler/rustc_resolve/src/late.rs

```diff
@@ -1752,13 +1896,208 @@ impl<'a: 'ast, 'b, 'ast> LateResolutionVisitor<'a, 'b, 'ast> {
+        let elision_failures = std::mem::replace(
+            &mut self.diagnostic_metadata.current_elision_failures,
+            outer_failures,
+        );
+        if let Err(failure_info) = elision_lifetime && !elision_failures.is_empty() {
+            self.report_missing_lifetime_specifiers(elision_failures, Some(failure_info));
+        }
+    }
+
+    fn elision_lifetime(
+        &mut self,
+        has_self: bool,
+        inputs: impl Iterator<Item = (Option<&'ast Pat>, &'ast Ty)>,
+    ) -> Result<LifetimeRes, (Vec<MissingLifetime>, Vec<ElisionFailureInfo>)> {
+        let outer_candidates =
+            std::mem::replace(&mut self.lifetime_elision_candidates, Some(Default::default()));
+
+        let mut elision_lifetime = None;
+        let mut lifetime_count = 0;
+        let mut parameter_info = Vec::new();
+
+        let mut bindings = smallvec![(PatBoundCtx::Product, Default::default())];
+        for (index, (pat, ty)) in inputs.enumerate() {
+            debug!(?pat, ?ty);
+            if let Some(pat) = pat {
+                self.resolve_pattern(pat, PatternSource::FnParam, &mut bindings);
+            }
+            self.visit_ty(ty);
+
+            if let Some(ref candidates) = self.lifetime_elision_candidates {
+                let new_count = candidates.len();
+                let local_count = new_count - lifetime_count;
+                if local_count != 0 {
+                    parameter_info.push(ElisionFailureInfo {
+                        index,
+                        ident: if let Some(pat) = pat && let PatKind::Ident(_, ident, _) = pat.kind {
+                            Some(ident)
+                        } else {
+                            None
+                        },
+                        lifetime_count: local_count,
+                        span: ty.span,
+                    });
+                }
+                lifetime_count = new_count;
+            }
+
+            // Handle `self` specially.
+            if index == 0 && has_self {
+                struct SelfVisitor<'r, 'a> {
+                    r: &'r mut Resolver<'a>,
+                    impl_self: Option<Res>,
+                    lifetime: Set1<LifetimeRes>,
+                }
+
+                impl SelfVisitor<'_, '_> {
+                    // Look for `self: &'a Self` - also desugared from `&'a self`,
+                    // and if that matches, use it for elision and return early.
+                    fn is_self_ty(&self, ty: &Ty) -> bool {
+                        match ty.kind {
+                            TyKind::ImplicitSelf => true,
+                            TyKind::Path(None, _) => {
+                                let path_res = self.r.partial_res_map[&ty.id].base_res();
+                                if let Res::SelfTy { .. } = path_res {
+                                    return true;
+                                }
+                                Some(path_res) == self.impl_self
```

Comment:
This logic looks extremely suspicious, but apparently it's an observable part of the language?
Do we have tests for this `TyKind::Path` case, when the path is not `Self` literally?

Reply:
Sigh.